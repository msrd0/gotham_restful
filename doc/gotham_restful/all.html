<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" 
        data-theme="ayu"
        data-use_system_theme="false"></script><script src="../storage.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../gotham_restful/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../gotham_restful/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location">Crate gotham_restful</h2></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../gotham_restful/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><h1 class="fqn"><span class="in-band">List of all items</span></h1><h3 id="Structs">Structs</h3><ul class="structs docblock"><li><a href="struct.AuthError.html">AuthError</a></li><li><a href="struct.AuthMiddleware.html">AuthMiddleware</a></li><li><a href="struct.NoContent.html">NoContent</a></li><li><a href="struct.NoopExtractor.html">NoopExtractor</a></li><li><a href="struct.OpenapiInfo.html">OpenapiInfo</a></li><li><a href="struct.Raw.html">Raw</a></li><li><a href="struct.Redirect.html">Redirect</a></li><li><a href="struct.Response.html">Response</a></li><li><a href="struct.StaticAuthHandler.html">StaticAuthHandler</a></li><li><a href="struct.Success.html">Success</a></li><li><a href="cors/struct.CorsConfig.html">cors::CorsConfig</a></li></ul><h3 id="Enums">Enums</h3><ul class="enums docblock"><li><a href="enum.AuthErrorOrOther.html">AuthErrorOrOther</a></li><li><a href="enum.AuthSource.html">AuthSource</a></li><li><a href="enum.AuthStatus.html">AuthStatus</a></li><li><a href="cors/enum.Headers.html">cors::Headers</a></li><li><a href="cors/enum.Origin.html">cors::Origin</a></li></ul><h3 id="Traits">Traits</h3><ul class="traits docblock"><li><a href="trait.AuthHandler.html">AuthHandler</a></li><li><a href="trait.DrawResourceRoutes.html">DrawResourceRoutes</a></li><li><a href="trait.DrawResourceRoutesWithSchema.html">DrawResourceRoutesWithSchema</a></li><li><a href="trait.DrawResources.html">DrawResources</a></li><li><a href="trait.DrawResourcesWithSchema.html">DrawResourcesWithSchema</a></li><li><a href="trait.Endpoint.html">Endpoint</a></li><li><a href="trait.EndpointWithSchema.html">EndpointWithSchema</a></li><li><a href="trait.FromBody.html">FromBody</a></li><li><a href="trait.GetOpenapi.html">GetOpenapi</a></li><li><a href="trait.IntoResponse.html">IntoResponse</a></li><li><a href="trait.IntoResponseError.html">IntoResponseError</a></li><li><a href="trait.IntoResponseWithSchema.html">IntoResponseWithSchema</a></li><li><a href="trait.RequestBody.html">RequestBody</a></li><li><a href="trait.Resource.html">Resource</a></li><li><a href="trait.ResourceWithSchema.html">ResourceWithSchema</a></li><li><a href="trait.ResponseBody.html">ResponseBody</a></li><li><a href="trait.ResponseSchema.html">ResponseSchema</a></li><li><a href="trait.WithOpenapi.html">WithOpenapi</a></li><li><a href="cors/trait.CorsRoute.html">cors::CorsRoute</a></li></ul><h3 id="Attribute-Macros">Attribute Macros</h3><ul class="attributes docblock"><li><a href="attr.create.html">create</a></li><li><a href="attr.delete.html">delete</a></li><li><a href="attr.delete_all.html">delete_all</a></li><li><a href="attr.endpoint.html">endpoint</a></li><li><a href="attr.read.html">read</a></li><li><a href="attr.read_all.html">read_all</a></li><li><a href="attr.search.html">search</a></li><li><a href="attr.update.html">update</a></li><li><a href="attr.update_all.html">update_all</a></li></ul><h3 id="Derive-Macros">Derive Macros</h3><ul class="derives docblock"><li><a href="derive.FromBody.html">FromBody</a></li><li><a href="derive.RequestBody.html">RequestBody</a></li><li><a href="derive.Resource.html">Resource</a></li><li><a href="derive.ResourceError.html">ResourceError</a></li></ul><h3 id="Functions">Functions</h3><ul class="functions docblock"><li><a href="cors/fn.handle_cors.html">cors::handle_cors</a></li></ul><h3 id="Typedefs">Typedefs</h3><ul class="typedefs docblock"><li><a href="type.AuthResult.html">AuthResult</a></li><li><a href="type.AuthSuccess.html">AuthSuccess</a></li><li><a href="type.AuthValidation.html">AuthValidation</a></li></ul></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="gotham_restful" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0 (897e37553 2022-11-02)" ></div></body></html>