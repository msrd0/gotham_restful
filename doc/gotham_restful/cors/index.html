<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `cors` mod in crate `gotham_restful`."><title>gotham_restful::cors - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-84e720fa.css"><script id="default-settings" 
data-use_system_theme="false"
data-theme="ayu"></script><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="gotham_restful" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0 (29483883e 2025-08-04)" data-channel="1.89.0" data-search-js="search-92309212.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../gotham_restful/index.html">gotham_<wbr>restful</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module cors</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate gotham_<wbr>restful</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">gotham_restful</a></div><h1>Module <span>cors</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/gotham_restful/cors.rs.html#1-299">Source</a> </span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CorsConfig.html" title="struct gotham_restful::cors::CorsConfig">Cors<wbr>Config</a></dt><dd>This is the configuration that the CORS handler will follow. Its default configuration is basically
not to touch any responses, resulting in the browser’s default behaviour.</dd></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.Headers.html" title="enum gotham_restful::cors::Headers">Headers</a></dt><dd>Specify the allowed headers of the request. It is up to the browser to check that only the allowed
headers are sent with the request.</dd><dt><a class="enum" href="enum.Origin.html" title="enum gotham_restful::cors::Origin">Origin</a></dt><dd>Specify the allowed origins of the request. It is up to the browser to check the validity of the
origin. This, when sent to the browser, will indicate whether or not the request’s origin was
allowed to make the request.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.CorsRoute.html" title="trait gotham_restful::cors::CorsRoute">Cors<wbr>Route</a></dt><dd>Add CORS routing for your path. This is required for handling preflight requests.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.handle_cors.html" title="fn gotham_restful::cors::handle_cors">handle_<wbr>cors</a></dt><dd>Handle CORS for a non-preflight request. This means manipulating the <code>res</code> HTTP headers so that
the response is aligned with the <code>state</code>’s <a href="struct.CorsConfig.html" title="struct gotham_restful::cors::CorsConfig">CorsConfig</a>.</dd></dl></section></div></main></body></html>