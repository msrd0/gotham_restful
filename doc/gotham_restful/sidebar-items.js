initSidebarItems({"attr":[["change",""],["change_all",""],["create",""],["endpoint",""],["read",""],["read_all",""],["remove",""],["remove_all",""],["search",""]],"derive":[["FromBody",""],["RequestBody",""],["Resource",""],["ResourceError",""]],"enum":[["AuthError","This is an error type that always yields a 403 Forbidden response. This type is best used in combination with [AuthSuccess] or [AuthResult]."],["AuthErrorOrOther","This is an error type that either yields a 403 Forbidden respone if produced from an authentication error, or delegates to another error type. This type is best used with [AuthResult]."],["AuthSource","The source of the authentication token in the request."],["AuthStatus","The authentication status returned by the auth middleware for each request."]],"mod":[["cors",""]],"struct":[["AuthMiddleware","This is the auth middleware. To use it, first make sure you have the `auth` feature enabled. Then simply add it to your pipeline and request it inside your handler:"],["AuthValidation","Contains the various validations that are applied after decoding a JWT."],["NoContent","This is the return type of a resource that doesn't actually return something. It will result in a 204 No Content answer by default. You don't need to use this type directly if using the function attributes:"],["NoopExtractor","A no-op extractor that can be used as a default type for [Endpoint::Placeholders] and [Endpoint::Params]."],["OpenapiInfo",""],["Raw","This type can be used both as a raw request body, as well as as a raw response. However, all types of request bodies are accepted by this type. It is therefore recommended to derive your own type from [RequestBody] and only use this when you need to return a raw response. This is a usage example that simply returns its body:"],["Redirect","This is the return type of a resource that only returns a redirect. It will result in a 303 See Other answer, meaning the redirect will always result in a GET request on the target."],["Response","A response, used to create the final gotham response from."],["StaticAuthHandler","An [AuthHandler] returning always the same secret. See [AuthMiddleware] for a usage example."],["Success","This can be returned from a resource when there is no cause of an error."]],"trait":[["AuthHandler","This trait will help the auth middleware to determine the validity of an authentication token."],["DrawResourceRoutes","This trait allows to draw routes within an resource. Use this only inside the [Resource::setup] method."],["DrawResourceRoutesWithSchema","This trait allows to draw routes within an resource. Use this only inside the [Resource::setup] method."],["DrawResources","This trait adds the `resource` method to gotham's routing. It allows you to register any RESTful [Resource] with a path."],["DrawResourcesWithSchema","This trait adds the `resource` method to gotham's routing. It allows you to register any RESTful [Resource] with a path."],["Endpoint",""],["EndpointWithSchema",""],["FromBody","This trait should be implemented for every type that can be built from an HTTP request body plus its media type."],["GetOpenapi","This trait adds the `get_openapi` and `swagger_ui` method to an OpenAPI-aware router."],["IntoResponse","This trait needs to be implemented by every type returned from an endpoint to to provide the response."],["IntoResponseError",""],["IntoResponseWithSchema","A trait provided to convert a resource's result to json, and provide an OpenAPI schema to the router. This trait is implemented for all types that implement [IntoResponse] and [ResponseSchema]."],["RequestBody","A type that can be used inside a request body. Implemented for every type that is deserializable with serde. If the `openapi` feature is used, it must also be of type OpenapiType."],["Resource","This trait must be implemented for every resource. It allows you to register the different endpoints that can be handled by this resource to be registered with the underlying router."],["ResourceType",""],["ResourceWithSchema","This trait must be implemented for every resource. It allows you to register the different endpoints that can be handled by this resource to be registered with the underlying router."],["ResponseBody","A type that can be used inside a response body. Implemented for every type that is serializable with serde. If the `openapi` feature is used, it must also be of type OpenapiType."],["ResponseSchema","Additional details for [IntoResponse] to be used with an OpenAPI-aware router."],["WithOpenapi","This trait adds the `with_openapi` method to gotham's routing. It turns the default router into one that will only allow RESTful resources, but record them and generate an OpenAPI specification on request."]],"type":[["AuthResult","This return type can be used to wrap any type implementing IntoResponse that can only be returned if the client is authenticated. Otherwise, an empty 403 Forbidden response will be issued."],["AuthSuccess","This return type can be used to wrap any type implementing IntoResponse that can only be returned if the client is authenticated. Otherwise, an empty 403 Forbidden response will be issued."]]});