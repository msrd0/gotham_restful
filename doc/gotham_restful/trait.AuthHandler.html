<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This trait will help the auth middleware to determine the validity of an authentication token."><title>AuthHandler in gotham_restful - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-42caa33d.css"><script id="default-settings" 
data-use_system_theme="false"
data-theme="ayu"></script><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="gotham_restful" data-themes="" data-resource-suffix="" data-rustdoc-version="1.84.1 (e71f9a9a9 2025-01-27)" data-channel="1.84.1" data-search-js="search-92e6798f.js" data-settings-js="settings-0f613d39.js" ><script src="../static.files/storage-59e33391.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-5f194d8c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../gotham_restful/index.html">gotham_<wbr>restful</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Auth<wbr>Handler</a></h2><h3><a href="#required-methods">Required Methods</a></h3><ul class="block"><li><a href="#tymethod.jwt_secret" title="jwt_secret">jwt_secret</a></li></ul><h3><a href="#dyn-compatibility">Dyn Compatibility</a></h3><h3><a href="#implementors">Implementors</a></h3></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate gotham_<wbr>restful</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><span class="rustdoc-breadcrumbs"><a href="index.html">gotham_restful</a></span><h1>Trait <span class="trait">AuthHandler</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/gotham_restful/auth.rs.html#102-109">Source</a> </span></div><pre class="rust item-decl"><code>pub trait AuthHandler&lt;Data&gt; {
    // Required method
    fn <a href="#tymethod.jwt_secret" class="fn">jwt_secret</a>&lt;F: <a class="trait" href="https://doc.rust-lang.org/1.84.1/core/ops/function/trait.FnOnce.html" title="trait core::ops::function::FnOnce">FnOnce</a>() -&gt; <a class="enum" href="https://doc.rust-lang.org/1.84.1/core/option/enum.Option.html" title="enum core::option::Option">Option</a>&lt;Data&gt;&gt;(
        &amp;self,
        state: &amp;mut <a class="struct" href="https://docs.rs/gotham/0.7.4/gotham/state/struct.State.html" title="struct gotham::state::State">State</a>,
        decode_data: F,
    ) -&gt; <a class="enum" href="https://doc.rust-lang.org/1.84.1/core/option/enum.Option.html" title="enum core::option::Option">Option</a>&lt;<a class="struct" href="https://doc.rust-lang.org/1.84.1/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec">Vec</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/1.84.1/std/primitive.u8.html">u8</a>&gt;&gt;;
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This trait will help the auth middleware to determine the validity of an authentication token.</p>
<p>A very basic implementation could look like this:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">const </span>SECRET: <span class="kw-2">&amp;</span><span class="lifetime">'static </span>[u8; <span class="number">32</span>] = <span class="string">b"zlBsA2QXnkmpe0QTh8uCvtAEa4j33YAc"</span>;

<span class="kw">struct </span>CustomAuthHandler;
<span class="kw">impl</span>&lt;T&gt; AuthHandler&lt;T&gt; <span class="kw">for </span>CustomAuthHandler {
	<span class="kw">fn </span>jwt_secret&lt;F: FnOnce() -&gt; <span class="prelude-ty">Option</span>&lt;T&gt;&gt;(
		<span class="kw-2">&amp;</span><span class="self">self</span>,
		_state: <span class="kw-2">&amp;mut </span>State,
		_decode_data: F
	) -&gt; <span class="prelude-ty">Option</span>&lt;Vec&lt;u8&gt;&gt; {
		<span class="prelude-val">Some</span>(SECRET.to_vec())
	}
}</code></pre></div>
</div></details><h2 id="required-methods" class="section-header">Required Methods<a href="#required-methods" class="anchor">ยง</a></h2><div class="methods"><details class="toggle method-toggle" open><summary><section id="tymethod.jwt_secret" class="method"><a class="src rightside" href="../src/gotham_restful/auth.rs.html#104-108">Source</a><h4 class="code-header">fn <a href="#tymethod.jwt_secret" class="fn">jwt_secret</a>&lt;F: <a class="trait" href="https://doc.rust-lang.org/1.84.1/core/ops/function/trait.FnOnce.html" title="trait core::ops::function::FnOnce">FnOnce</a>() -&gt; <a class="enum" href="https://doc.rust-lang.org/1.84.1/core/option/enum.Option.html" title="enum core::option::Option">Option</a>&lt;Data&gt;&gt;(
    &amp;self,
    state: &amp;mut <a class="struct" href="https://docs.rs/gotham/0.7.4/gotham/state/struct.State.html" title="struct gotham::state::State">State</a>,
    decode_data: F,
) -&gt; <a class="enum" href="https://doc.rust-lang.org/1.84.1/core/option/enum.Option.html" title="enum core::option::Option">Option</a>&lt;<a class="struct" href="https://doc.rust-lang.org/1.84.1/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec">Vec</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/1.84.1/std/primitive.u8.html">u8</a>&gt;&gt;</h4></section></summary><div class="docblock"><p>Return the SHA256-HMAC secret used to verify the JWT token.</p>
</div></details></div><h2 id="dyn-compatibility" class="section-header">Dyn Compatibility<a href="#dyn-compatibility" class="anchor">ยง</a></h2><div class="dyn-compatibility-info"><p>This trait is <b>not</b> <a href="https://doc.rust-lang.org/1.84.1/reference/items/traits.html#object-safety">dyn compatible</a>.</p><p><i>In older versions of Rust, dyn compatibility was called "object safety", so this trait is not object safe.</i></p></div><h2 id="implementors" class="section-header">Implementors<a href="#implementors" class="anchor">ยง</a></h2><div id="implementors-list"><section id="impl-AuthHandler%3CT%3E-for-StaticAuthHandler" class="impl"><a class="src rightside" href="../src/gotham_restful/auth.rs.html#127-135">Source</a><a href="#impl-AuthHandler%3CT%3E-for-StaticAuthHandler" class="anchor">ยง</a><h3 class="code-header">impl&lt;T&gt; <a class="trait" href="trait.AuthHandler.html" title="trait gotham_restful::AuthHandler">AuthHandler</a>&lt;T&gt; for <a class="struct" href="struct.StaticAuthHandler.html" title="struct gotham_restful::StaticAuthHandler">StaticAuthHandler</a></h3></section></div><script src="../trait.impl/gotham_restful/auth/trait.AuthHandler.js" async></script></section></div></main></body></html>