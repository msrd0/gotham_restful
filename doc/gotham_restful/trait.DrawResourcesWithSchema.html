<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This trait adds the `resource` method to gotham’s routing. It allows you to register any RESTful Resource with a path."><title>DrawResourcesWithSchema in gotham_restful - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-492a78a4a87dcc01.css"><script id="default-settings" 
data-theme="ayu"
data-use_system_theme="false"></script><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="gotham_restful" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0 (f6e511eec 2024-10-15)" data-channel="1.82.0" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../gotham_restful/index.html">gotham_<wbr>restful</a><span class="version">0.9.0</span></h2></div><h2 class="location"><a href="#">Draw<wbr>Resources<wbr>With<wbr>Schema</a></h2><div class="sidebar-elems"><section><h3><a href="#required-methods">Required Methods</a></h3><ul class="block"><li><a href="#tymethod.resource">resource</a></li></ul><h3><a href="#object-safety">Object Safety</a></h3><h3><a href="#implementors">Implementors</a></h3></section><h2><a href="index.html">In crate gotham_<wbr>restful</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Trait <a href="index.html">gotham_restful</a>::<wbr><a class="trait" href="#">DrawResourcesWithSchema</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/gotham_restful/routing.rs.html#46">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>pub trait DrawResourcesWithSchema {
    // Required method
    fn <a href="#tymethod.resource" class="fn">resource</a>&lt;R: <a class="trait" href="trait.ResourceWithSchema.html" title="trait gotham_restful::ResourceWithSchema">ResourceWithSchema</a>&gt;(&amp;mut self, path: &amp;<a class="primitive" href="https://doc.rust-lang.org/1.82.0/std/primitive.str.html">str</a>);
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This trait adds the <code>resource</code> method to gotham’s routing. It allows you to register
any RESTful <a href="trait.Resource.html" title="trait gotham_restful::Resource">Resource</a> with a path.</p>
</div></details><h2 id="required-methods" class="section-header">Required Methods<a href="#required-methods" class="anchor">§</a></h2><div class="methods"><section id="tymethod.resource" class="method"><a class="src rightside" href="../src/gotham_restful/routing.rs.html#50">source</a><h4 class="code-header">fn <a href="#tymethod.resource" class="fn">resource</a>&lt;R: <a class="trait" href="trait.ResourceWithSchema.html" title="trait gotham_restful::ResourceWithSchema">ResourceWithSchema</a>&gt;(&amp;mut self, path: &amp;<a class="primitive" href="https://doc.rust-lang.org/1.82.0/std/primitive.str.html">str</a>)</h4></section></div><h2 id="object-safety" class="section-header">Object Safety<a href="#object-safety" class="anchor">§</a></h2><div class="object-safety-info">This trait is <b>not</b> <a href="https://doc.rust-lang.org/1.82.0/reference/items/traits.html#object-safety">object safe</a>.</div><h2 id="implementors" class="section-header">Implementors<a href="#implementors" class="anchor">§</a></h2><div id="implementors-list"></div><script src="../trait.impl/gotham_restful/routing/trait.DrawResourcesWithSchema.js" async></script></section></div></main></body></html>