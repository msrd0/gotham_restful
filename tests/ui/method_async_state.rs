#[macro_use] extern crate gotham_restful;
use gotham_restful::State;

#[derive(Resource)]
#[resource(read_all)]
struct FooResource;

#[read_all(FooResource)]
async fn read_all(state : &State)
{
}

fn main()
{
}
