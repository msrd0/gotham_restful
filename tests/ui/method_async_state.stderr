error: async fn must not take &State as an argument as State is not Sync, consider taking &mut State
 --> $DIR/method_async_state.rs:9:19
  |
9 | async fn read_all(state : &State)
  |                   ^^^^^

error[E0433]: failed to resolve: use of undeclared type or module `_gotham_restful_resource_foo_resource_method_read_all`
 --> $DIR/method_async_state.rs:4:10
  |
4 | #[derive(Resource)]
  |          ^^^^^^^^ use of undeclared type or module `_gotham_restful_resource_foo_resource_method_read_all`
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused import: `gotham_restful::State`
 --> $DIR/method_async_state.rs:2:5
  |
2 | use gotham_restful::State;
  |     ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
